staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,581 @@
+						// remove the request
+						req.Queue.Requests = append(req.Queue.Requests[:i],
+							req.Queue.Requests[i+1:]...)
+						break
+					}
+				}
+				// At this point, if the qs is quiescing,
+				// has zero requests executing, and has zero requests enqueued
+				// then a call to the EmptyHandler should be forked.
+				if qs.emptyHandler != nil && qs.numRequestsEnqueued == 0 &&
+					qs.getRequestsExecutingLocked() == 0 {
+					qs.counter.Add(1)
+					go func(eh fq.EmptyHandler) {
+						defer qs.counter.Add(-1)
+						defer runtime.HandleCrash()
+						eh.HandleEmpty()
+					}(qs.emptyHandler)
+				}
+			} else {
+				// At this point we know that req was in its queue earlier and another
+				// goroutine has removed req from its queue and called qs.counter.Add(1)
+				// in anticipation of unblocking this goroutine through the other arm of this
+				// select. In this case we need to decrement the counter because this goroutine
+				// was actually unblocked through a different code path.
+				qs.counter.Add(-1)
+			}
+		}()
+		return false, false, func() {}
+	}
+}
+
+// syncTimeLocked is used to sync the time of the queueSet by looking at the elapsed
+// time since the last sync and this value based on the 'virtualtime ratio'
+// which scales inversely to the # of active flows
+func (qs *queueSet) syncTimeLocked() {
+	realNow := qs.clock.Now()
+	timesincelast := realNow.Sub(qs.lastRealTime).Seconds()
+	qs.lastRealTime = realNow
+	var virtualTimeRatio float64
+
+	activeQueues := 0
+	reqs := 0
+	for _, queue := range qs.queues {
+		reqs += queue.RequestsExecuting
+
+		if len(queue.Requests) > 0 || queue.RequestsExecuting > 0 {
+			activeQueues++
+		}
+	}
+	if activeQueues != 0 {
+		// TODO(aaron-prindle) document the math.Min usage
+		virtualTimeRatio = math.Min(float64(reqs), float64(qs.config.ConcurrencyLimit)) / float64(activeQueues)
+	}
+
+	qs.virtualTime += timesincelast * virtualTimeRatio
+}
+
+func (qs *queueSet) lockAndSyncTime() {
+	qs.lock.Lock()
+	qs.syncTimeLocked()
+}
+
+// timeoutOldRequestsAndRejectOrEnqueueLocked encapsulates the logic required
+// to validate and enqueue a request for the queueSet/QueueSet:
+// 1) Start with shuffle sharding, to pick a queue.
+// 2) Reject old requests that have been waiting too long
+// 3) Reject current request if there is not enough concurrency shares and
+// we are at max queue length
+// 4) If not rejected, create a request and enqueue
+// returns the enqueud request on a successful enqueue
+// returns nil in the case that there is no available concurrency or
+// the queuelengthlimit has been reached
+func (qs *queueSet) timeoutOldRequestsAndRejectOrEnqueueLocked(hashValue uint64) *fq.Request {
+	//	Start with the shuffle sharding, to pick a queue.
+	queueIdx := qs.chooseQueueIndexLocked(hashValue)
+	queue := qs.queues[queueIdx]
+	// The next step is the logic to reject requests that have been waiting too long
+	qs.removeTimedOutRequestsFromQueue(queue)
```

Comment:
It is deliberately done in this order so that the shuffle sharding sees the queues prior to the removal of the detritus? If that's the case, the comment on the function should say so and explain why this is desirable.

I guess I would have expected all the queues to be processed for this prior to running the shuffle sharding, so that it has an accurate count of what is in each queue.